### DISCUSSION TITLE
Rust fltk-rs scrollbar increment issue

### DISCUSSION THREAD

**Discussion by fiddler55** [fltk-rs/fltk-rs#1586] [Q&A] [answered]
Scrollbars typically have 2 types of increments using a mouse, a "line" increment and a "block" increment. A line increment is done by clicking on either arrow at the ends of the scrollbar, while a block increment is done by clicking on either side of the slider. I have a scrollbar that I want to represent a range from 0 to 360 (degrees), with a line increment of 1 (degree), and a block increment of 15 (degrees). How do I do this with the fltk toolkit? I have the following sample Rust program:

![Screenshot from 2024-09-21 14-19-25](https://github.com/user-attachments/assets/1e13d698-2f47-498c-a3d6-8eef7b393911)

```
// Testing the fltk scrollbar
// Be sure to have the following line in Cargo.toml after [dependencies]:
// fltk = { version = "1.4.15", features = ["fltk-bundled"] }
use fltk::{prelude::*, window::Window, *};

fn main() {
    let app = app::App::default();
    let mut wind = Window::default()
        .with_size(400, 100)
        .with_label("Scrollbar test");

    let mut col = group::Flex::default_fill().column();
    col.set_pad(0);

    let mut text = frame::Frame::default().with_label("0");

    let mut bar = valuator::Scrollbar::default()
        .with_type(valuator::ScrollbarType::Horizontal);
    bar.set_range(0.0, 360.0);
    bar.set_value(0.0);
    // bar.set_step(<f64>, <i32>);
    col.fixed(&bar, 20);

    col.end();

    wind.end();
    wind.show();

    bar.set_callback({
        move |s| {
            text.set_label(&(s.value().round() as i32).to_string());
            s.parent().unwrap().redraw();
        }
    });

    app.run().unwrap();
}

```

I tried using different parameters for bar.set_step(...), but could not get the line increment to be different than block increment, and there seems to be a built-in default of 16. Is there something else I should try? Desiring to have a line increment of 1, and a block increment of 15. Thanks in advance for any help.

**Comment by MoAlyousef** [comment]
Fl_Scrollbar has a linesize method which wasn't exposed in fltk-rs. 
I've added Scrollbar::linesize() and Scrollbar::set_linesize(i32) in the master branch.
Can you check if that works for you using:
```toml
[dependencies]
fltk = "^1.4"

[patch.crates-io]
fltk = { git = "https://github.com/fltk-rs/fltk-rs" }
```

**Reply by fiddler55** [reply]
MoAlyousef,

Thanks, I used bar.set_linesize(1) and now clicking on the arrows moves the slider by 1. But I can't get the "block" increment (i.e. clicking on either side of the slider) to be 15. It stays at 1. What would I do to get the "line" increment to stay at 1 and the "block" increment to be 15?

**Reply by MoAlyousef** [reply]
I checked FLTK's `Fl_Scrollbar::increment_cb()` and it appears that pushing (arrows or around the slider box) only changes by linesize.
Dragging actually changes by step.
It appears it might be desirable to change pushing around the slider box by step.
I'll post a question in the fltk mailing list and see what answer I can get.

**Reply by fiddler55** [reply]
The scrollbar represents rotation (0 - 360 degrees).
Just to make sure you understand what I'm looking for:
1) When clicking either arrow I want the slider to move by 1 (degree). I assume this is the "linesize".
2) When dragging the slider it would be good if it can be set to any increment (0 - 360 degrees, by degree). I assume this is what "stepsize" would mean, and thus I would want it to be 1 (so I can set it to any degree).
3) When clicking on either side of the slider I want the slider to move by 15 (degrees). This is want I termed "blocksize". (It appears the slider moves by 16, 1 more than I want. 15 is desirable since it evenly divides into 360.)
This is a common behavior I've seen in scrollbars.

**Comment by MoAlyousef** [comment]
I've opened a discussion a few days ago at the FLTK mailing list here:
https://groups.google.com/g/fltkgeneral/c/N1pPvc6qLTg

From previous discussion I don't think the default behavior will change before FLTK cuts a 1.4 stable release.

**Reply by MoAlyousef** [reply]
Hi @fiddler55 

Can you try:
```
    let mut bar = valuator::Scrollbar::default()
        .with_type(valuator::ScrollbarType::Horizontal);
    bar.scroll_value(0, 15, 0, 360);
    bar.set_linesize(1);
    col.fixed(&bar, 20);
```
with the current master branch?

**Comment by fiddler55** [comment]
I get:

```
error[E0599]: no method named `scroll_value` found for struct `Scrollbar` in the current scope
  --> src/main.rs:23:9
   |
19 |     bar.scroll_value(0, 15, 0, 360);
   |         ^^^^^^^^^^^^ method not found in `Scrollbar`

```

By the way, is there a distinction between set_bounds and set_range?

**Reply by MoAlyousef** [reply]
You would need to pull the latest master. Modify your Cargo.toml with:
```toml
[dependencies]
fltk = "^1.4"

[patch.crates-io]
fltk = { git = "https://github.com/fltk-rs/fltk-rs" }
```
Then run `cargo update && cargo run`

**Reply by fiddler55** [reply]
Oops, did cargo clean, then cargo run. Didn't do cargo update. Now I'm slightly wiser...

bar.scroll_value(0, 15, 0, 360);  // Gave me what I wanted, a 15 degree increment. Unfortunately the scrollbar only went from 0 to 345. So I tried:

bar.scroll_value(0, 15, 0, 375);  // This went from 0 to 360, but with a 14 degree increment. Then I tried:

bar.scroll_value(0, 16, 0, 375);  // This went from 0 to 359 with a 15 degree increment. Finally:

bar.scroll_value(0, 16, 0, 376);  // This went from 0 to 360 with a 15 degree increment, which is what I wanted. Success!?

Not sure how this works from the numbers entered - Is there a bug in the code?

**Reply by MoAlyousef** [reply, marked as answer]
I noticed that setting the range should work:
```rust
    bar.scroll_value(0, 15, 0, 360);
    bar.set_range(0.0, 360.0);
    bar.set_linesize(1);
```

**Reply by fiddler55** [reply]
This works! Thanks Mohammed!
